### ISSUE TITLE
Is location of `HTMLClose` comment correct?

### ISSUE THREAD

**Issue by fisker** [meriyah/meriyah#154]
```js
comments = []; require('meriyah').parse('<!--\na\n-->', {onComment: comments, webcompat: true, ranges: true});comments
[
  { type: 'HTMLOpen', value: '', start: 0, end: 4, range: [ 0, 4 ] },
  { type: 'HTMLClose', value: '', start: 6, end: 10, range: [ 6, 10 ] }
]
```

We have `[6, 10]`, while babel has `[7, 10]`

```js
require('@babel/parser').parse('<!--\na\n-->').comments
[
  {
    type: 'CommentLine',
    value: '',
    start: 0,
    end: 4,
    loc: SourceLocation {
      start: [Position],
      end: [Position],
      filename: undefined,
      identifierName: undefined
    }
  },
  {
    type: 'CommentLine',
    value: '',
    start: 7,
    end: 10,
    loc: SourceLocation {
      start: [Position],
      end: [Position],
      filename: undefined,
      identifierName: undefined
    }
  }
]
```

**Comment by 3cp** [comment]
That's what I read from ecma, SingleLineHTMLCloseComment starts with the line terminator?

http://www.ecma-international.org/ecma-262/11.0/index.html#prod-annexB-SingleLineHTMLCloseComment

Babel seems doing for "HTMLCloseComment". But based on the spec, I think it should use SingleLineHTMLCloseComment, as that's parallel to SingleLineHTMLOpenComment which has no corresponding HTMLOpenComment.

```
SingleLineHTMLOpenComment::
  <!-- SingleLineCommentCharsopt
SingleLineHTMLCloseComment::
  LineTerminatorSequence HTMLCloseComment
```

Basically, the spec said there is no HTMLClose comment if you do `<!-- lorem --> lorem2` (it's one whole HTMLOpen comment).

`\n-->` or `-->` on first line is the start of a HTMLClose comment.

**Comment by fisker** [comment]
I guess the range in `meriyah` make more sense, closing.

**Comment by 3cp** [comment]
Just curious, is there any real world code using html comments in js file? Or this is just for completeness of prettier?

**Comment by fisker** [comment]
I don't think there is any real world code today, but it could be in some old html file. We enabled `webcompat` to parse `"\8"`, so there will be possible to have it. Not big deal.

**Comment by fisker** [comment]
I don't think there is any real world code today, but it could be in some old html file. We enabled `webcompat` to parse `"\8"`, so there will be possible to have it. Not big deal.

**Comment by KFlash** [comment]
@3cp Should Meriyah implement another algorithm for comment extracting? Meriyah extract the comments the same way as Acorn, but in the end it turns out that the comment algorithm for inserting comments isn't perfect. Babel, Acorn, and Esprima suffers from it. No one seem to get it right.

@fisker What is the biggest known issues with injecting comments in Prettier? Trailing? Leading?

I went down this rabbit hole myself therefore I started to wonder when I saw all the different issues on the mentioned repos.

**Comment by fisker** [comment]
Prettier use own comment attach logic, but many problems, for example, some time we move comments out `()` and people prefer keep it at the original location.  And some comment can't be attached, example `[, , /* comment */ , ]` there is no node there element is null. 

I was thinking add something like `printCommentsBetweenTokenOrNode`. All parsers has token info.

**Comment by KFlash** [comment]
Do you have a list over all the comment issues and comment that can be attached, like your  `[, , /* comment */ , ]` example?

How would you develop `printCommentsBetweenTokenOrNode` ?

Let me check if my own emitter can print that "impossible" comment ;)

**Comment by KFlash** [comment]
@fisker I tried that "impossible" comment with my own comment attach logic, and it seems to work. Need to make it more "pretty", but it got attached

![works](https://user-images.githubusercontent.com/31855118/98895234-9c181e80-24e1-11eb-8161-f5537bbcc758.png)

**Comment by fisker** [comment]
> How would you develop printCommentsBetweenTokenOrNode ?

Example

```js
for (/* comment 1 */;/* comment 2 */;/* comment 3 */);
```

Logic

```js

if (!node.init) {
printCommentsBetweenTokenOrNode('(', ';')
}
if (!node.test) {
printCommentsBetweenTokenOrNode(';', ';')
}
if (!node.update) {
printCommentsBetweenTokenOrNode(';', ')')
}
```

**Comment by fisker** [comment]
> Do you have a list over all the comment issues and comment that can be attached, like your [, , /* comment */ , ] example?

Just `null` in Node properties.

We use `Node.comments` to store comments.

**Comment by KFlash** [comment]
Your solution seems expensive? For this comment `for (/* comment 1 */;/* comment 2 */;/* comment 3 */);`. 
Calculate the spaces between `(` and `)`. Then use a code that can grab the comments in that range. You also need to calculate on leading and trailing. 

The point is too only attach the comments, right? So after you got that array returned, you attach the comment and use the range and source length and there you go ! :)

**Comment by fisker** [comment]
Another example,

```js
const foo = new Foo/* comment*/;
```

If we decide to add `()` to new call.

Should we format it to


```js
const foo = new Foo(/* comment*/);
```

or

```js
const foo = new Foo()/* comment*/;
```

**Comment by KFlash** [comment]
Is there a place we can chat outside a comment section, and I can help you too solve the comment issues permanently. I'm working on it right now in my own code.

**Comment by KFlash** [comment]
Regarding your last examples. It depends on the original source I guess. If it's written with or without parens. But I would suggest that - if no parent to start with - it should belong to 'Foo'. So if you add parens it should be inside, not outside. 

It all depends on the original source. I read from right to left when it comes to comments. That way you will always have a "parent" on left side to attach too. If not well then you sit back with a leading comment like `// foo`

**Comment by fisker** [comment]
DM me https://twitter.com/fisker

**Comment by KFlash** [comment]
I tweeted and mentioned you there. Anyway. Check your email. After that you can go to the /emitter folder where I'm playing around ATM.  And also check out last part of the whitespace.ts file in the scanner folder.

I think that should answer your problems
